Move `fpsemi-examples`

MTWhyte opened the pull request:
This (work-in-progress) PR moves the `fpsemi-examples` functions out of `tests` and into `include/libsemigroups` / `src`, the idea being to make them accessible to `libsemigroups_pybind11`.

Aside from changing the directories of the relevant files, this PR:

- documents the functions in `fpsemi-examples.hpp`,
- changes some functions which formerly had initials in their names to use `author` values,
- standardises each construction to end with `monoid` or `semigroup`, and use snake case,
- adds co-authors to the formerly `author::Guralnick` presentation for the symmetric group.
- adds references to presentations where possible
- specifies whether the presentation returned is a semigroup or monoid presentation

MTWhyte commented:
I’ll reach the point of considering tests for these this afternoon.

@james-d-mitchell, do you think it makes the most sense to just collect existing tests using these functions into a new file? For example, taking the tests which compute the sizes of various semigroups out of `test-todd-coxeter` and similar, and into a new file `test-fp-semigroups`.

I think this seems sensible to do where possible, but possibly there are drawbacks I’ve not considered.

MTWhyte commented:
I combined the bulk of the changes into one commit. Since this PR adds a lot to the `libsemigroups.bib` file, I thought I'd just run the whole file through a formatter so it looks uniform. I've also deleted the test which doesn't actually test what it should, of issue #404. Both of these smaller changes are done in separate commits to make it easy to drop them or put them in a different PR.

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,537 @@
+  }
+
+}  // namespace libsemigroups
+#endif  // LIBSEMIGROUPS_TESTS_FPSEMI_EXAMPLES_HPP_
```
```suggestion
#endif  // LIBSEMIGROUPS_FPSEMI_EXAMPLES_HPP_
```

MTWhyte commented:
I've added some of the missing references in the way we discussed. I've also added a paper of Donald Knuth to the bibliography for the plactic monoid (he called it the tableau algebra). The link in the documentation page still points to a different paper, which calls it the plactic monoid, and mentions this fact.

james-d-mitchell commented on Makefile.am:
```diff
@@ -832,9 +832,12 @@ test_felsch_tree_SOURCES =  tests/test-felsch-tree.cpp
-test_fpsemi_SOURCES += tests/fpsemi-examples.cpp
 test_fpsemi_SOURCES += tests/test-main.cpp
 
+test_fpsemi_examples_SOURCES = tests/test-fpsemi-examples-1.cpp
```
```suggestion
test_fpsemi_examples_SOURCES  = tests/test-fpsemi-examples-1.cpp
```

james-d-mitchell commented on benchmarks/bench-sims1.cpp:
```diff
@@ -33,6 +33,7 @@
 #include "libsemigroups/cong-wrap.hpp"  // for CongruenceWrapper
 #include "libsemigroups/constants.hpp"  // for PositiveInfinity, POSITIVE...
 #include "libsemigroups/fastest-bmat.hpp"
+#include "libsemigroups/fpsemi-examples.hpp"
```
```suggestion
#include "libsemigroups/fpsemi-examples.hpp" // for ???
```

james-d-mitchell commented on docs/source/api/fpsemi-examples.rst:
```diff
@@ -0,0 +1,154 @@
+.. Copyright (c) 2022, J. D. Mitchell
```
```suggestion
.. Copyright (c) 2022, M. T. Whyte
```
Me thinks

james-d-mitchell commented on docs/source/api/fpsemi-examples.rst:
```diff
@@ -0,0 +1,154 @@
+
+.. cpp:namespace:: libsemigroups
+
+Examples
+--------------------
```
```suggestion
Examples
----------
```

james-d-mitchell commented on docs/source/api/fpsemi-examples.rst:
```diff
@@ -0,0 +1,154 @@
+Examples
+--------------------
+
+Defined in ``present.hpp`` and ``knuth-bendix.hpp``.
```
I'm not sure that this line or the next paragraph is really what you meant, looks more like it's copy pasta ;)

james-d-mitchell commented on docs/source/api/fpsemi-examples.rst:
```diff
@@ -0,0 +1,154 @@
+manipulating :cpp:type:`Presentation` objects. All such functions are contained in
+the namespace ``presentation``.
+
+These functions are available in the ``present.hpp`` file except
```
Same for this paragraph.

james-d-mitchell commented on docs/source/api/fpsemi-examples.rst:
```diff
@@ -0,0 +1,154 @@
+These functions are available in the ``present.hpp`` file except
+:cpp:any:`redundant_rule` which is in ``knuth-bendix.hpp``.
+
+Presentations from the following sources are implemented: :cite:`Gay1999aa`; :cite:`Godelle2009aa`; :cite:`East2022aa`; :cite:`Maltcev2007aa`;
```
Can you please reflow this so that is uses only 80 chars per line? Thanks!

james-d-mitchell commented on docs/source/api/fpsemi-examples.rst:
```diff
@@ -0,0 +1,154 @@
+   :header-rows: 0
+
+   * - :cpp:any:`symmetric_group`
+     - Construct a presentation for the symmetric group.
```
```suggestion
     - Presentation for the symmetric group.
```
Not sure if this is any better, but maybe remove the "Construct a" from each line here?

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+//
+// libsemigroups - C++ library for semigroups and monoids
+// Copyright (C) 2019 Florent Hivert
```
```suggestion
// Copyright (C) 2022 Murray T. Whyte
```

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+#include "libsemigroups/types.hpp"    // for relation_type
+
+namespace libsemigroups {
+  enum class author : uint64_t {
```
This should be documented too, see for example, 

https://libsemigroups.readthedocs.io/en/latest/types.html

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+    Sutov      = 65536
+  };
+
+  inline author operator+(author auth1, author auth2) {
```
This should probably be documented too

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+                               + static_cast<uint64_t>(auth2));
+  }
+
+  std::vector<relation_type> rook_monoid(size_t l, int q);
```
Please add a comment or TODO about why this isn't documented.

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  // \noexcept
+  std::vector<relation_type> renner_common_type_B_monoid(size_t l, int q);
+
+  std::vector<relation_type> RennerTypeBMonoid(size_t l, int q);
```
Add comment or TODO about why not documented

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  // The argument `val` determines the specific presentation
+  // which is returned. The options are:
+  // * `author::Godelle` (described
+  // in [10.48550/arXiv.0904.0926](https://doi.org/10.48550/arXiv.0904.0926),
```
You might consider using `[10.48550/arXiv.0904.0926][]` and collecting the links at the end of this doc like:

[10.48550/arXiv.0904.0926]: https://doi.org/10.48550/arXiv.0904.0926

james-d-mitchell replied:
[10.48550/arXiv.0904.0926]: https://whatever.xyz

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  // `author::Godelle`.
+  std::vector<relation_type> renner_type_D_monoid(size_t l, int q, author val);
+
+  std::vector<relation_type> RennerTypeDMonoid(size_t l, int q);
```
Same comment or TODO about why no doc

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //! \returns A `std::vector<relation_type>`
+  //!
+  //! \throws LibsemigroupsException if `n < 3`, or the author value is not
+  //! `author::Easdown + author::East + author::FitzGerald`.
```
```suggestion
```
There's no author argument here!

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //! stellar monoid with `n` generators, as in Theorem 4.39 of
+  //! [10.48550/arXiv.1910.11740](https://doi.org/10.48550/arXiv.1910.11740).
+  //!
+  //! \param n the number of generators
```
```suggestion
  //! \param l the number of generators
```
Should use the same names for parameters as in the actual function signature.

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //!
+  //! \returns A `std::vector<relation_type>`
+  //!
+  //! \throws LibsemigroupsException if `n < 3`, or the author value is not
```
Same here either `l` or `n` not both

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //!
+  //! \returns A `std::vector<relation_type>`
+  //!
+  //! \throws LibsemigroupsException if `n < 3`, or the author value is not
```
```suggestion
  //! \throws LibsemigroupsException if `n < 3`
  //! \throws LibsemigroupsException if the `author` is not one of those listed above.
```

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //!
+  //! \returns A `std::vector<relation_type>`
+  //!
+  //! \noexcept
```
Really?

james-d-mitchell replied:
Are you sure this function returns the right answer when `n` is `0` for example?

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //! as described in
+  //! [10.1007/s10012-000-0001-1](https://doi.org/10.1007/s10012-000-0001-1).
+  //!
+  //! \param n the order
```
```suggestion
  //! \param n the order of the chain
```

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //! Fibonacci semigroup \f$F(r, n)\f$, as described in
+  //! [10.1016/0022-4049(94)90005-1](https://doi.org/10.1016/0022-4049(94)90005-1).
+  //!
+  //! \param r the parameter `r`
```
```suggestion
  //! \param r the length of the left hand sides of the relations.
```

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //!
+  //! \returns A `std::vector<relation_type>`
+  //!
+  //! \noexcept
```
Not sure I believe this

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //!
+  //! \returns A `std::vector<relation_type>`
+  //!
+  //! \throws LibsemigroupsException if `val` is not listed above (modulo order
```
separate `\throws` for separate failing conditions please

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //! [http://hdl.handle.net/10023/2821](http://hdl.handle.net/10023/2821))
+  //! * `author::Iwahori` (see Theorem 9.3.1 of
+  //! [10.1007/978-1-84800-281-4](https://doi.org/10.1007/978-1-84800-281-4))
+  //!
```
```suggestion
```

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  //!
+  //! \returns A `std::vector<relation_type>`
+  //!
+  //! \throws LibsemigroupsException if `val` is not listed above (modulo order
```
Not sure I believe this exactly either :(

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  }
+
+  template <typename T>
+  T make_group(T& M) {
```
I think this whole function should be deleted.

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+  }
+
+  template <typename T>
+  T AlternatingGroupMoore(size_t n) {
```
This function should be renamed `alternating_group` and documented + behave like all the other functions in this file.

james-d-mitchell commented on src/fpsemi-examples.cpp:
```diff
@@ -621,135 +634,145 @@ namespace libsemigroups {
-        return 2;
-      }
-    };
+      LIBSEMIGROUPS_EXCEPTION("expected first argument to be at least 3");
```
```suggestion
      LIBSEMIGROUPS_EXCEPTION("expected the 1st argument (size_t) to be at least 3, found %llu", uint64_t(n));
```

james-d-mitchell commented on src/fpsemi-examples.cpp:
```diff
@@ -621,135 +634,145 @@ namespace libsemigroups {
+      }
       return result;
+    } else {
+      LIBSEMIGROUPS_EXCEPTION("not yet implemented");
```
Maybe it'd be most helpful to re-iterate what the possible valid argument combinations are here?

james-d-mitchell commented on tests/test-fpsemi-examples-2.cpp:
```diff
@@ -0,0 +1,89 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+// This file is the seonc of two containing tests for the fpsemi-examples
```
```suggestion
// This file is the second ? of two containing tests for the fpsemi-examples
```

james-d-mitchell commented on include/libsemigroups/fpsemi-examples.hpp:
```diff
@@ -0,0 +1,539 @@
+
+  template <typename T,
+            typename
+            = std::enable_if_t<std::is_base_of<PresentationBase, T>::value>>
```
```suggestion
            = std::enable_if_t<std::is_same<Presentation<word_type>, T>::value>>
```
As we discussed just now, this actually doesn't work for `std::string` so we should exclude that for now.

MTWhyte commented:
I will also add default author values, and mention in the documentation what these are.

MTWhyte commented:
I noticed there was no test using the Aizenstat presentation for the full transformation monoid, so went to add one.
```
    LIBSEMIGROUPS_TEST_CASE("fpsemi-examples",
                            "038",
                            "full_transformation_monoid(4) Aizenstat",
                            "[fpsemi-examples][quick]") {
      auto   rg = ReportGuard(true);
      auto   s  = full_transformation_monoid(4, author::Aizenstat);
      auto   p  = make<Presentation<word_type>>(s);
      p.alphabet(5);
      presentation::replace_word(p, word_type({}), {4});
      presentation::add_identity_rules(p, 4);
      ToddCoxeter tc(congruence_kind::twosided);
      tc.set_number_of_generators(5);
      for (size_t i = 0; i < p.rules.size() - 1; i += 2) {
        tc.add_pair(p.rules[i], p.rules[i + 1]);
      }
      REQUIRE(tc.number_of_classes() == 256);
    }
```

This ran for a concerningly long time before I stopped it. Getting a printout of `p.rules`, I get the following:

```
{ { 0, 0 }, { 4 },
{ 1, 1, 1, 1 }, { 4 },
{ 0, 1, 0, 1, 0, 1 }, { 4 },
{ 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1 }, { 4 },
{ 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1 }, { 4 },
{ 1, 3 }, { 3 },
{ 2, 2, 1, 2, 2, 3, 2, 2, 1, 2, 2 }, { 3 },
{ 2, 1, 2, 2, 2, 1, 2, 3, 2, 2, 2, 1, 2, 1, 2, 2, 2 }, { 3 },
{ 3, 2, 1, 2, 2, 2, 3, 2, 1, 2, 2, 2 }, { 3 },
{ 2, 2, 2, 1, 2, 3, 2, 2, 2, 1, 2, 3 }, { 3, 2, 2, 2, 1, 2, 3 },
{ 3, 2, 2, 2, 1, 2, 3, 2, 2, 2, 1, 2 }, { 3, 2, 2, 2, 1, 2, 3 },
{ 3, 2, 1, 2, 2, 1, 2, 3, 2, 1, 2, 2, 1, 2 }, { 2, 1, 2, 2, 1, 2, 3, 2, 1, 2, 2, 1, 2, 3 },
{ 0, 4 }, { 0 },
{ 4, 0 }, { 0 },
{ 1, 4 }, { 1 },
{ 4, 1 }, { 1 },
{ 2, 4 }, { 2 },
{ 4, 2}, { 2 },
{ 3, 4 }, { 3 },
{ 4, 3 }, { 3 },
{ 4, 4 }, { 4 } }
```

Here, `4` is the identity. Checking this against the source,
<img width="547" alt="image" src="https://user-images.githubusercontent.com/42549861/203077216-539fb07c-c45d-4914-8983-57ffc997cb9f.png">
everything looks as it should to me. I wonder if perhaps the Aizenstat presentation is another candidate for the 'non-examples'? I think what I will do for the meantime is just remove it from the documentation, to finish off this PR.

james-d-mitchell commented:
Looks like some of the tests need the [no-valgrind] tag @MTWhyte
